get.go

```diff
@@ -391,51 +395,126 @@ func resolvePackage(
 				}
 			}
 
-			// In this case rely on output parsing.
+			// In this case it is not successful from our perspective.
 			gerr = errors.New(out)
 		}
 	}
 
-	if verbose {
-		logger.Println("tricky: Matching go get error output:", gerr.Error())
+	// We fallback only if go-get failed which happens when it does not know what version to choose.
+	// In this case
+	if err := resolveInGoModCache(logger, verbose, update, target); err != nil {
+		return errors.Wrapf(err, "fallback to local go mod cache resolution failed after go get failure: %v", gerr)
 	}
+	return nil
+}
 
-	// TODO(bwplotka) Obviously hacky but reliable so far.
-	// Try to match strings announcing what version was found (if we got to this stage).
-	downloadingRe := fmt.Sprintf(`go: downloading (%v) (\S*)`, target.Path())
-	upgradeRe := `go: (\S*) upgrade => (\S*)`
-	foundVersionRe := fmt.Sprintf(`go: found %v in (\S*) (\S*)`, target.Path())
+func gomodcache() string {
+	cachepath := os.Getenv("GOMODCACHE")
+	if gpath := os.Getenv("GOPATH"); gpath != "" && cachepath == "" {
+		cachepath = filepath.Join(gpath, "pkg/mod")
+	}
+	return cachepath
+}
 
-	re, err := regexp.Compile(downloadingRe)
+func latestModVersion(listFile string) (_ string, err error) {
+	f, err := os.Open(listFile)
 	if err != nil {
-		return errors.Wrapf(err, "regexp compile %v", downloadingRe)
+		return "", err
 	}
-	if !re.MatchString(gerr.Error()) {
-		re = regexp.MustCompile(upgradeRe)
-		if !re.MatchString(gerr.Error()) {
-			re, err = regexp.Compile(foundVersionRe)
+	defer errcapture.Do(&err, f.Close, "list file close")
+
+	scanner := bufio.NewScanner(f)
+	var lastVersion string
+	for scanner.Scan() {
+		lastVersion = scanner.Text()
+	}
+	if err := scanner.Err(); err != nil {
+		return "", err
+	}
+	if lastVersion == "" {
+		return "", errors.New("empty file")
+	}
+	return lastVersion, nil
```

Comment:
If I'm not wrong, this returns the last version found in the file, which seems to be how the file is laid out. Given that the `semver` library already has functions to compare, is it worth sorting the versions based on semver instead of returning the last one?

Reply:
I don't think this will work, since there are those custom versions too (with sha and date). But maybe it follows semver (: Then yes.

In practice what I checked this file is sorted, but you are right it would be better to ensure this just in case as this behavior is not stated anywhere. Let's do in separate PR